modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTask.java

```diff
@@ -298,6 +300,81 @@ private static ElasticsearchException wrapException(Exception ex) {
         return storedContext;
     }
 
+    private static class Assigner {
+
+        private final ReindexIndexClient reindexIndexClient;
+        private final String persistentTaskId;
+        private final long allocationId;
+        private int assignmentAttempts = 0;
+
+        private Assigner(ReindexIndexClient reindexIndexClient, String persistentTaskId, long allocationId) {
+            this.reindexIndexClient = reindexIndexClient;
+            this.persistentTaskId = persistentTaskId;
+            this.allocationId = allocationId;
+        }
+
+        private void assign(Listener assignmentListener) {
+            ++assignmentAttempts;
+            reindexIndexClient.getReindexTaskDoc(persistentTaskId, new ActionListener<>() {
+                @Override
+                public void onResponse(ReindexTaskState taskState) {
+                    long term = taskState.getPrimaryTerm();
+                    long seqNo = taskState.getSeqNo();
+                    ReindexTaskStateDoc oldDoc = taskState.getStateDoc();
+                    ReindexRequest request = oldDoc.getReindexRequest();
+                    BulkByScrollResponse response = oldDoc.getReindexResponse();
+                    ElasticsearchException exception = oldDoc.getException();
+                    RestStatus failureStatusCode = oldDoc.getFailureStatusCode();
+                    ScrollableHitSource.Checkpoint checkpoint = oldDoc.getCheckpoint();
+
+                    if (oldDoc.getAllocationId() == null || allocationId > oldDoc.getAllocationId()) {
+                        ReindexTaskStateDoc newDoc = new ReindexTaskStateDoc(request, allocationId, response, exception, failureStatusCode,
+                            checkpoint);
+                        reindexIndexClient.updateReindexTaskDoc(persistentTaskId, newDoc, term, seqNo, new ActionListener<>() {
+                            @Override
+                            public void onResponse(ReindexTaskState newTaskState) {
+                                assignmentListener.onAssignment(newTaskState);
+                            }
+
+                            @Override
+                            public void onFailure(Exception ex) {
+                                if (ex instanceof VersionConflictEngineException) {
+                                    // There has been an indexing operation since the GET operation. Try
+                                    // again if there are assignment attempts left.
+                                    if (assignmentAttempts < 3) {
```

Comment:
I added a TODO. I think that I still need to think through and implement the cancelled() functionality (where an external signal indicates cancel). This will require stoping cancelling the updating and reindex process. Which feels like a separate task.